Oracle Database Embedding Store

Michael-A-McMahon opened the pull request:
<!-- Thank you so much for your contribution! -->
<!-- Please fill in all the sections below. -->

<!-- Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples. -->
<!-- Please note that PRs with breaking changes will be rejected. -->
<!-- Please note that PRs without tests will be rejected. -->

<!-- Please note that PRs will be reviewed based on the priority of the issues they address. -->
<!-- We ask for your patience. We are doing our best to review your PR as quickly as possible. -->
<!-- Please refrain from pinging and asking when it will be reviewed. Thank you for understanding! -->


## Issue
Closes #1091 

## Change
Added an `EmbeddingStore` integration for Oracle Database.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new model integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [X] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)


## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

psilberk committed:
First commit

psilberk committed:
Removed forced dependency to TestContainers

psilberk committed:
Merge pull request #1 from psilberk/jdbc_vector_store

Jdbc vector store

psilberk committed:
Refactor sample

psilberk committed:
Merge pull request #2 from psilberk/jdbc_vector_store

Refactor sample

Michael-A-McMahon committed:
Made some optimizations

Signed-off-by: Michael McMahon <michael.a.mcmahon@oracle.com>

Michael-A-McMahon committed:
Merge pull request #3 from psilberk/optimizations

Made some optimizations

Michael-A-McMahon committed:
New implementation

Michael-A-McMahon committed:
Completed SQLFilter implementation

Michael-A-McMahon committed:
Added tests for removal

Michael-A-McMahon committed:
Constructing SQLFilter with key mapper

Michael-A-McMahon committed:
Typos and search break

Michael-A-McMahon committed:
Merge pull request #4 from psilberk/fresh-start

Fresh start

Michael-A-McMahon committed:
Added DistanceMetric

Michael-A-McMahon committed:
Renamed module

Michael-A-McMahon committed:
Merge pull request #11 from psilberk/oracle-bom-update

Module Rename

Michael-A-McMahon commented:
This PR adds a basic foundation for integration with Oracle Database as an embedding store.

We plan to enhance this integration with future work:
- Configurable vector index parameters
- Configurable distance metrics
- Creation of a JSON index for metadata filtering 

Please note that we are working on refactoring the new tests to use Test Containers (https://testcontainers.com/). I'm planning to push a new commit with that change sometime today.

For now, the code is ready for review. Please let me know what you think!

Michael-A-McMahon commented on langchain4j-bom/pom.xml:
```diff
@@ -257,6 +257,12 @@
 
+            <dependency>
+                <groupId>dev.langchain4j</groupId>
+                <artifactId>langchain4j-oracle-database</artifactId>
```
We should discuss naming. Originally, we thought to use the short "langchain4j-oracle" name. But then we thought Oracle makes many products, some of which might also have a place in langchain4j. The Generative AI service of Oracle Cloud comes to mind. So to distinguish which product we are integrating with, we thought to add "database" into the name.

But maybe we are over thinking it :)

Please let me know what you think. I'm happy to make a change if you think the short name works better. Just wanted to lay out our rationale.

Michael-A-McMahon replied:
Makes sense. I'll rename to just langchain4j-oracle.

Michael-A-McMahon commented on langchain4j-oracle-database/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java:
```diff
@@ -0,0 +1,589 @@
+ * supports metadata filtering and removal. Instances of this embedding store are safe for use by multiple threads.
+ * </p><p>
+ * Instances of this store are created by configuring a builder:
+ * </p><pre>{@code
```
Not required. It's just a convention I took from OpenJDK. Example: https://github.com/openjdk/jdk/blob/c25c4896ad9ef031e3cddec493aef66ff87c48a7/src/java.base/share/classes/java/util/stream/Stream.java#L54
If you view the JavaDoc in IntelliJ, it adds syntax highlighting to the code example, which I think is nice.

But I can go with a simple `<pre>` if you'd like. Please let me know.

Michael-A-McMahon commented on langchain4j-oracle-database/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java:
```diff
@@ -0,0 +1,589 @@
+        this.dataSource = ensureNotNull(builder.dataSource, "dataSource");
+        this.tableName = ensureNotNull(builder.tableName, "tableName");
+
+        createSchema(dataSource, tableName);
```
We use an `IF NOT EXISTS` clause in our `CREATE TABLE` SQL. This means we don't get an error if the table already exists.
Or is the `createTable` option something that _recreates_ the table (and index), if they already exist? Or is the idea to just avoid an extra SQL call on creation of the embedding store?

Michael-A-McMahon replied:
I really appreciate this comment. I think we definitely need to design for the case where tables and indexes are already created, and the user does not have privileges to create or recreate them.

Michael-A-McMahon replied:
Definitely want the default to be no table or index creation.

I was thinking there could be at least 3 desirable behaviors:
```java
/** An option for how a schema object, such as a table, is created. */
enum CreateOption {

   /** No attempt is made to create the schema object. */
   NONE,

  /** An existing schema object is reused, otherwise it is created */
  IF_NOT_EXISTS,

  /** An existing schema object is dropped and replaced with a new one */
  CREATE_OR_REPLACE

}
```
With the enum, we don't need:
```java
OracleEmbeddingStore.builder()
  .dropExistingTable(true)
  .createTable(true)
  .dropExistingIndex(true)
  .createIndex(true)
  ...
```
Instead:
```java
OracleEmbeddingStore.builder()
  .tableCreation(CREATE_OR_REPLACE)
  .indexCreation(CREATE_OR_REPLACE)
  ...
```
The enum makes the code a bit more expressive and readable.
An enum also means we can add new options in the future, without cluttering-up the builder interface with new methods. 
Not sure how much of an issue it would be, but the enum does prevent invalid combinations of builder calls, such as:
```java
OracleEmbeddingStore.builder()
  .dropExistingTable(true)
  .createTable(false)
  ...
```

Michael-A-McMahon replied:
We added a `CreateOption` enum. The API ended like this:
```java
OracleEmbeddingStore.builder()
  .embeddingTable("my_embedding_table", CreateOption.CREATE_IF_NOT_EXISTS)
  .vectorIndex(CreateOption.CREATE_OR_REPLACE)
```
The default is no table or index creation.

Michael-A-McMahon committed:
Documenting test config

Michael-A-McMahon committed:
Fixed JavaDocs and added test config

Michael-A-McMahon committed:
Delete commented code

Michael-A-McMahon committed:
Remove unused variable

Michael-A-McMahon committed:
Merge branch 'main' of github.com:psilberk/langchain4j into configurable-distance

Michael-A-McMahon committed:
Support exact/approximate search

Michael-A-McMahon committed:
Skip index creation for exact search

Michael-A-McMahon committed:
Merge pull request #12 from psilberk/test-envvars

Test Containers and Environment Variables

Michael-A-McMahon committed:
Added configurable column names

Michael-A-McMahon committed:
Added test for column names

Michael-A-McMahon committed:
Add unicode test

Michael-A-McMahon committed:
Added EmbeddingTable

Michael-A-McMahon committed:
Merge pull request #13 from psilberk/configurable-distance

Distance Metrics and Exact Search

Michael-A-McMahon committed:
Starting JavaDoc fixes

Michael-A-McMahon committed:
Merge branch 'main' of github.com:psilberk/langchain4j into configurable-names

Michael-A-McMahon committed:
Updated JavaDocs and used CommonTestOperations

Michael-A-McMahon committed:
Merge pull request #14 from psilberk/configurable-names

Configurable Column names

psilberk committed:
Refactor folder

psilberk committed:
Merge pull request #15 from psilberk/refactor-folder

Refactor folder

psilberk committed:
Adding UCP for tests

psilberk committed:
Changing connection factory class

psilberk committed:
Merge pull request #16 from psilberk/add-ucp-tests

Adding UCP for tests

Michael-A-McMahon committed:
Added Oracle embedding store

psilberk committed:
Merge pull request #17 from psilberk/integrations-doc

Added Oracle embedding store

Michael-A-McMahon committed:
Minor fixes

psilberk committed:
Merge pull request #18 from psilberk/sqlfilter-not-public

Minor Fixes

Michael-A-McMahon committed:
Added VectorIndex class

Michael-A-McMahon committed:
Added Builder.vectorIndex(CreateOption0

psilberk committed:
Merge pull request #19 from psilberk/minimal-vector-index

Vector Index Create Option

Michael-A-McMahon commented on langchain4j-oracle-database/README.md:
````diff
@@ -0,0 +1,16 @@
+```xml
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artificatId>langchain4j-oracle</artificatId>
````
Changed to `langchain4j-oracle` in the latest commit.

Michael-A-McMahon commented on langchain4j-oracle-database/pom.xml:
Added these in the latest commit today. Verified by building with Docusaurus and viewing in a web browser.

Michael-A-McMahon commented on langchain4j-oracle-database/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java:
```diff
@@ -0,0 +1,589 @@
+ * {@link IllegalStateException#getCause()} when a database operation results in an error.
+ * </p>
+ */
+public final class OracleEmbeddingStore implements EmbeddingStore<TextSegment> {
```
We added an `EmbeddingTable` object with a builder that configures non-default column names. We can even have non-English characters, if the names are wrapped in double-quotes!

Michael-A-McMahon commented on langchain4j-oracle-database/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java:
```diff
@@ -0,0 +1,589 @@
+                        UUID uuid = UUID.fromString(string);
+                        metadata.put(key, uuid);
+                    }
+                    catch (IllegalArgumentException notUUID) {
```
Latest commit puts JSON String values as String entries into the Metadata object. No more UUID parsing.

Michael-A-McMahon commented on langchain4j-oracle-database/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java:
```diff
@@ -0,0 +1,589 @@
+     * @return
+     */
+    private static RuntimeException uncheckSQLException(SQLException sqlException) {
+        return new IllegalStateException(sqlException);
```
Latest commit has replaced IllegalStateException with RuntimeException.

Michael-A-McMahon commented on langchain4j-oracle-database/src/test/java/dev/langchain4j/store/embedding/oracle/TestDataSource.java:
```diff
@@ -0,0 +1,118 @@
+ * {@link #getConnection()} will return the same connection on successive calls. This allows tests to complete much
+ * faster than they would if a new connection was created on each call. More sophisticated connection pools are readily
+ * available, but using this pool will avoid introducing a new dependency. Not introducing dependencies is one of the
+ * guidelines listed in the CONTRIBUTING.md document.
```
Latest commit has tests using Universal Connection Pool (UCP), and my little TestDataSource is removed. My team develops UCP in tandem with the Oracle JDBC Driver, so it should be a good option; We are constantly running integration tests to ensure that UCP is compatible with the latest builds of our driver.

Michael-A-McMahon commented on langchain4j-oracle-database/src/main/java/dev/langchain4j/store/embedding/oracle/SQLFilter.java:
```diff
@@ -0,0 +1,90 @@
+ * }</pre>
+ * </p>
+ */
+public interface SQLFilter {
```
Fixed.

Michael-A-McMahon committed:
Update docs/docs/integrations/embedding-stores/oracle.md

Set version to 0.34.0

Co-authored-by: LangChain4j <langchain4j@gmail.com>

Michael-A-McMahon committed:
Update langchain4j-oracle/pom.xml

Set version to 0.34.0

Co-authored-by: LangChain4j <langchain4j@gmail.com>

Michael-A-McMahon commented on langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java:
```diff
@@ -0,0 +1,693 @@
+                while (resultSet.next()) {
+
+                    // Convert the cosine distance between 0 and 2 into a score between 1 and 0.
+                    double score = 1d - (resultSet.getDouble("distance") / 2d);
```
I'm very glad you noticed this, because I completely missed it!

I'd like to get your opinion: Do you think that supporting distance metrics, other than cosine, makes sense? I ask this question in the context of langchain4j, where the EmbeddingMatch.score() method is specified as:
```java
    /**
     * Returns the relevance score (derivative of cosine distance) of this embedding compared to
     * a reference embedding during a search.
     * The current implementation assumes that the embedding store uses cosine distance when comparing embeddings.
     *
     * @return Relevance score, ranging from 0 (not relevant) to 1 (highly relevant).
     */
```
Based on this, it seems that an implementation of `search(EmbeddingRequest)` is only correct if it uses cosine distance. So, if we allow users to configure a different metric, then we are allowing them to create an instance of EmbeddingStore which does not behave correctly.

Do you think we should remove the configurable distance metric, and just to stick to cosine?

Michael-A-McMahon replied:
I think the best move is to remove support for non-cosine metrics for now. It keeps the user experience simpler, and it keeps our code simpler. Once the API is ready, we can add the support back, and we can do it in a way that is consistent with all other embedding stores.

Michael-A-McMahon commented:
> Hi, I am trying to run tests, but getting this error:
> 
> ```
> java.lang.ExceptionInInitializerError
> 	at dev.langchain4j.store.embedding.oracle.OracleEmbeddingStoreWithFilteringIT.<clinit>(OracleEmbeddingStoreWithFilteringIT.java:13)
> 	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
> 	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
> 	at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:549)
> 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
> 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
> 	at java.base/java.util.Optional.orElseGet(Optional.java:364)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> Caused by: org.testcontainers.containers.ContainerLaunchException: Container startup failed for image gvenzl/oracle-free:23.4-slim-faststart
> 	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:359)
> 	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:330)
> 	at dev.langchain4j.store.embedding.oracle.CommonTestOperations.<clinit>(CommonTestOperations.java:75)
> 	... 9 more
> Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
> 	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
> 	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:344)
> 	... 11 more
> Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
> 	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:563)
> 	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:354)
> 	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
> 	... 12 more
> Caused by: org.testcontainers.containers.ContainerLaunchException: Timed out waiting for log output matching '.*DATABASE IS READY TO USE!.*\s'
> 	at org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy.waitUntilReady(LogMessageWaitStrategy.java:47)
> 	at org.testcontainers.containers.wait.strategy.AbstractWaitStrategy.waitUntilReady(AbstractWaitStrategy.java:52)
> 	at org.testcontainers.oracle.OracleContainer.waitUntilContainerStarted(OracleContainer.java:81)
> 	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:500)
> 	... 14 more
> ```
> 
> I've tried to add `.withStartupTimeoutSeconds(600).withConnectTimeoutSeconds(600)` to
> 
> ```java
> new OracleContainer("gvenzl/oracle-free:23.4-slim-faststart")
>                     .withDatabaseName("pdb1")
>                     .withUsername("testuser")
>                     .withPassword("testpwd")
> ```
> 
> , but it does not seem to help.
> 
> Docker container logs:
> 
> ```
> CONTAINER: starting up...
> CONTAINER: first database startup, initializing...
> CONTAINER: starting up Oracle Database...
> LSNRCTL for Linux: Version 23.0.0.0.0 - Production on 19-AUG-2024 12:54:51
> Copyright (c) 1991, 2024, Oracle.  All rights reserved.
> Starting /opt/oracle/product/23ai/dbhomeFree/bin/tnslsnr: please wait...
> TNSLSNR for Linux: Version 23.0.0.0.0 - Production
> System parameter file is /opt/oracle/product/23ai/dbhomeFree/network/admin/listener.ora
> Log messages written to /opt/oracle/diag/tnslsnr/8d39556eca81/listener/alert/log.xml
> Listening on: (DESCRIPTION=(ADDRESS=(PROTOCOL=ipc)(KEY=EXTPROC_FOR_FREE)))
> Listening on: (DESCRIPTION=(ADDRESS=(PROTOCOL=tcp)(HOST=0.0.0.0)(PORT=1521)))
> Connecting to (DESCRIPTION=(ADDRESS=(PROTOCOL=IPC)(KEY=EXTPROC_FOR_FREE)))
> STATUS of the LISTENER
> ------------------------
> Alias                     LISTENER
> Version                   TNSLSNR for Linux: Version 23.0.0.0.0 - Production
> Start Date                19-AUG-2024 12:54:52
> Uptime                    0 days 0 hr. 0 min. 0 sec
> Trace Level               off
> Security                  ON: Local OS Authentication
> SNMP                      OFF
> Default Service           FREE
> Listener Parameter File   /opt/oracle/product/23ai/dbhomeFree/network/admin/listener.ora
> Listener Log File         /opt/oracle/diag/tnslsnr/8d39556eca81/listener/alert/log.xml
> Listening Endpoints Summary...
>   (DESCRIPTION=(ADDRESS=(PROTOCOL=ipc)(KEY=EXTPROC_FOR_FREE)))
>   (DESCRIPTION=(ADDRESS=(PROTOCOL=tcp)(HOST=0.0.0.0)(PORT=1521)))
> The listener supports no services
> The command completed successfully
> ORACLE instance started.
> Total System Global Area 1603726344 bytes
> Fixed Size                  5360648 bytes
> Variable Size             788529152 bytes
> Database Buffers          805306368 bytes
> Redo Buffers                4530176 bytes
> Database mounted.
> ```
> 
> Do you happen to know what is the problem? Thanks!

I'm asking around with my team, hopefully someone knows. I personally don't use TestContainers as the docker image doesn't seem to work on my Mac M1. 
In case it's helpful, the README.md lists some envvars that will override TestContainers, and have tests connect to your own Oracle DB. This is what I use for testing.

fmeheust commented:
> > Hi, I am trying to run tests, but getting this error:
> > ```
> > java.lang.ExceptionInInitializerError
> > 	at dev.langchain4j.store.embedding.oracle.OracleEmbeddingStoreWithFilteringIT.<clinit>(OracleEmbeddingStoreWithFilteringIT.java:13)
> > 	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
> > 	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
> > 	at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:549)
> > 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
> > 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
> > 	at java.base/java.util.Optional.orElseGet(Optional.java:364)
> > 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> > 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> > Caused by: org.testcontainers.containers.ContainerLaunchException: Container startup failed for image gvenzl/oracle-free:23.4-slim-faststart
> > 	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:359)
> > 	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:330)
> > 	at dev.langchain4j.store.embedding.oracle.CommonTestOperations.<clinit>(CommonTestOperations.java:75)
> > 	... 9 more
> > Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
> > 	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
> > 	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:344)
> > 	... 11 more
> > Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
> > 	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:563)
> > 	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:354)
> > 	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
> > 	... 12 more
> > Caused by: org.testcontainers.containers.ContainerLaunchException: Timed out waiting for log output matching '.*DATABASE IS READY TO USE!.*\s'
> > 	at org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy.waitUntilReady(LogMessageWaitStrategy.java:47)
> > 	at org.testcontainers.containers.wait.strategy.AbstractWaitStrategy.waitUntilReady(AbstractWaitStrategy.java:52)
> > 	at org.testcontainers.oracle.OracleContainer.waitUntilContainerStarted(OracleContainer.java:81)
> > 	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:500)
> > 	... 14 more
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I've tried to add `.withStartupTimeoutSeconds(600).withConnectTimeoutSeconds(600)` to
> > ```java
> > new OracleContainer("gvenzl/oracle-free:23.4-slim-faststart")
> >                     .withDatabaseName("pdb1")
> >                     .withUsername("testuser")
> >                     .withPassword("testpwd")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > , but it does not seem to help.
> > Docker container logs:
> > ```
> > CONTAINER: starting up...
> > CONTAINER: first database startup, initializing...
> > CONTAINER: starting up Oracle Database...
> > LSNRCTL for Linux: Version 23.0.0.0.0 - Production on 19-AUG-2024 12:54:51
> > Copyright (c) 1991, 2024, Oracle.  All rights reserved.
> > Starting /opt/oracle/product/23ai/dbhomeFree/bin/tnslsnr: please wait...
> > TNSLSNR for Linux: Version 23.0.0.0.0 - Production
> > System parameter file is /opt/oracle/product/23ai/dbhomeFree/network/admin/listener.ora
> > Log messages written to /opt/oracle/diag/tnslsnr/8d39556eca81/listener/alert/log.xml
> > Listening on: (DESCRIPTION=(ADDRESS=(PROTOCOL=ipc)(KEY=EXTPROC_FOR_FREE)))
> > Listening on: (DESCRIPTION=(ADDRESS=(PROTOCOL=tcp)(HOST=0.0.0.0)(PORT=1521)))
> > Connecting to (DESCRIPTION=(ADDRESS=(PROTOCOL=IPC)(KEY=EXTPROC_FOR_FREE)))
> > STATUS of the LISTENER
> > ------------------------
> > Alias                     LISTENER
> > Version                   TNSLSNR for Linux: Version 23.0.0.0.0 - Production
> > Start Date                19-AUG-2024 12:54:52
> > Uptime                    0 days 0 hr. 0 min. 0 sec
> > Trace Level               off
> > Security                  ON: Local OS Authentication
> > SNMP                      OFF
> > Default Service           FREE
> > Listener Parameter File   /opt/oracle/product/23ai/dbhomeFree/network/admin/listener.ora
> > Listener Log File         /opt/oracle/diag/tnslsnr/8d39556eca81/listener/alert/log.xml
> > Listening Endpoints Summary...
> >   (DESCRIPTION=(ADDRESS=(PROTOCOL=ipc)(KEY=EXTPROC_FOR_FREE)))
> >   (DESCRIPTION=(ADDRESS=(PROTOCOL=tcp)(HOST=0.0.0.0)(PORT=1521)))
> > The listener supports no services
> > The command completed successfully
> > ORACLE instance started.
> > Total System Global Area 1603726344 bytes
> > Fixed Size                  5360648 bytes
> > Variable Size             788529152 bytes
> > Database Buffers          805306368 bytes
> > Redo Buffers                4530176 bytes
> > Database mounted.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Do you happen to know what is the problem? Thanks!
> 
> I'm asking around with my team, hopefully someone knows. I personally don't use TestContainers as the docker image doesn't seem to work on my Mac M1. In case it's helpful, the README.md lists some envvars that will override TestContainers, and have tests connect to your own Oracle DB. This is what I use for testing.

I am running on a Windows machine and for now I do not manage to reproduce the error. I will look into it, try to reproduce and fix.

dliubarskyi committed:
Merge branch 'main' into main

fmeheust commented:
> Thank you for looking into the problem. While I can test with Oracle installed locally, it is not the best long-term solution and it would be great to have it working with docker

Are you also running on a Mac ? I have found some information in the image : https://github.com/gvenzl/oci-oracle-free/blob/main/README.md#oracle-database-free-on-apple-m-chips.

fmeheust commented:
> @fmeheust No, Windows

@langchain4j ARM ?

Michael-A-McMahon committed:
Removed configurable distance metric feature

Michael-A-McMahon committed:
Merge branch 'langchain4j:main' into main

Michael-A-McMahon committed:
Updated pom for new ojdbc and langchain4j sync

psilberk committed:
Merge pull request #24 from psilberk/pom-update

Updated pom for new ojdbc and langchain4j sync

fmeheust committed:
OracleContainer workaround to set timeout. Timeout set to 5mins.

fmeheust committed:
Increased timeout to 10mins

fmeheust committed:
Set waitStrategy startup timeout when startup timeout is changed.

psilberk committed:
Merge pull request #25 from psilberk/test-containers-workaround

OracleContainer workaround to set timeout. Timeout set to 5mins.

fmeheust commented:
We have found an issue in OracleContainers that does not change the startup timeout of the wait strategy when the timeout is set. Since the wait strategy is initialized on the constructor it is always set to 60s not matter what is set using withStartupTimeoutSeconds(int). We will submit a patch for TestContainers, but since it is probably going to take some time until it is merged and released, we have also added a workaround to this pull request. We have created a subclass of OracleContainer that will correctly set the timeout.

@langchain4j I have tested with two different Windows machines and I did not manage to reproduce. Can you please verify that the 10 minute timeout I have set fixes the problem for you?

fmeheust commented:
> @fmeheust thanks a lot, it works now!
> 
> But now some tests fail:
> 
> ![image](https://private-user-images.githubusercontent.com/132277850/359799986-70dc3557-7743-463d-b34c-7e961f296de4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQyMjY2NjcsIm5iZiI6MTcyNDIyNjM2NywicGF0aCI6Ii8xMzIyNzc4NTAvMzU5Nzk5OTg2LTcwZGMzNTU3LTc3NDMtNDYzZC1iMzRjLTdlOTYxZjI5NmRlNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyMVQwNzQ2MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZjQxM2IxYTZlNzk5MGJhMjdjNDljODg5MjVjODM2MTg1YmI1MDZhMWJjMWQ2NzFiZGM4Yzg3MjM1NDgxZTkyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.n2X9lMax7x3wTHe-o9skX-upCm3VzNGBpccI6iP5Ke8)
> 
> Most of them fail due to the error you can see in the screenshot, only `testUnicodeNames` has different error:
> 
> > Caused by: Error : 942, Position : 72, SQL = CREATE VECTOR INDEX IF NOT EXISTS "δεδομένα_embedding_index" ON "δεδομένα"("埋め込み") ORGANIZATION NEIGHBOR PARTITIONS WITH DISTANCE COSINE, Original SQL = CREATE VECTOR INDEX IF NOT EXISTS "δεδομένα_embedding_index" ON "δεδομένα"("埋め込み") ORGANIZATION NEIGHBOR PARTITIONS WITH DISTANCE COSINE, Error Message = ORA-00942: table or view "TESTUSER"."ΔΕΔΟΜΈΝΑ" does not exist

@langchain4j I manage to reproduce this error in my personal machine which runs Windows in French. Do you mind telling me what your local setting are?

fmeheust committed:
Add warning to the log when an error occurs while creating the vector index. The store can be used without an index.

fmeheust committed:
Fixed indentation

fmeheust committed:
Removed lines that should not be there, used for testing.

Michael-A-McMahon commented:
An Oracle database can be configured to recognize any character as a decimal point for numbers. This effect comparisons between NUMBER and VARCHAR, as the VARCHAR is parsed into a NUMBER.
So I think the DB is setup with something other than `.` as the decimal point, and raises the error when it sees that character in a VARCHAR.

Reference:
https://docs.oracle.com/en/database/oracle/oracle-database/23/refrn/NLS_NUMERIC_CHARACTERS.html
https://docs.oracle.com/en/database/oracle/oracle-database/23/refrn/NLS_TERRITORY.html

I'll need to think about how to fix this...

psilberk committed:
Merge pull request #26 from psilberk/create-vector-index-error

Add warning to the log when an error occurs while creating the vector index. The store can be used without an index.

eddumelendez commented on langchain4j-oracle/src/test/java/dev/langchain4j/store/embedding/oracle/CommonTestOperations.java:
```diff
@@ -0,0 +1,219 @@
+                DATA_SOURCE.setURL(oracleContainer.getJdbcUrl());
+                DATA_SOURCE.setUser(oracleContainer.getUsername());
+                DATA_SOURCE.setPassword(oracleContainer.getPassword());
+            }
```
There is no need to create `OracleContainerWithTimeout`. There is an issue with `withStartupTimeoutSeconds` that is tracked in https://github.com/testcontainers/testcontainers-java/issues/8590 but `withStartupTimeout(Duration.ofSeconds(600))` can be used.

```suggestion
                OracleContainer oracleContainer =
                    new OracleContainer("gvenzl/oracle-free:23.4-slim-faststart")
                        .withStartupTimeout(Duration.ofSeconds(600))
                        .withConnectTimeoutSeconds(600)
                        .withDatabaseName("pdb1")
                        .withUsername("testuser")
                        .withPassword("testpwd");
                oracleContainer.start();

                DATA_SOURCE.setURL(oracleContainer.getJdbcUrl());
                DATA_SOURCE.setUser(oracleContainer.getUsername());
                DATA_SOURCE.setPassword(oracleContainer.getPassword());
            }
```

fmeheust replied:
Thank you!

fmeheust replied:
That suggested changes are not enough, we should also remove the OracleContainerWithTimeout class.

Michael-A-McMahon committed:
Added SQLType in SQLFilters

psilberk committed:
Merge pull request #27 from psilberk/filter-sql-type

Explicit SQL types for SQLFilter

dliubarskyi committed:
Update langchain4j-oracle/src/test/java/dev/langchain4j/store/embedding/oracle/CommonTestOperations.java

Co-authored-by: Eddú Meléndez Gonzales <eddu.melendez@gmail.com>

fmeheust committed:
Removed OracleContainerWithTimeout and added missing import (#28)

Merged directly without review because the missing import causes test class not to build.

psilberk committed:
Merge pull request #23 from psilberk/remove-distance-metric

Removed configurable distance metric feature

fmeheust committed:
Added usage section to README

fmeheust committed:
Added links

fmeheust committed:
Fixed link

psilberk committed:
Merge pull request #29 from psilberk/update-readme

Update readme

fmeheust committed:
Minor changes

psilberk committed:
Merge pull request #30 from psilberk/minor-comments

Minor changes

fmeheust commented on langchain4j-oracle/README.md:
````diff
@@ -13,6 +13,69 @@ This module implements `EmbeddingStore` using Oracle Database.
 </dependency>
 ```
 
+## Usage
````
I have added it to both files

fmeheust commented on langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java:
```diff
@@ -134,12 +133,15 @@ private static void createIndex(Builder builder) throws SQLException {
 
             statement.executeBatch();
+        } catch (SQLException sqlException) {
+            log.warn("Error creating index: " + sqlException.getMessage(), sqlException);
```
Yes it should, the user explicitly asked for the index to be created, the exception should be propagated

dliubarskyi committed:
Update langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/SQLFilters.java

dliubarskyi committed:
Merge branch 'main' into main

jeandelavarene commented:
Thank you @langchain4j for your support!
